LFS-1041, LFS-1042: As a user entering answers to a multi-valued question, I am warned that using "," or ";" to separate values in the input does not automatically generate separate answer entries

marta- opened the pull request:
Also includes[ LFS-1042: _Some cardiac_rehab questions should have separator detection enabled_](https://phenotips.atlassian.net/browse/LFS-1042).

**To test**: `cardiac_rehab` run mode > `Baseline Health Information` > `Allergies`, and enter comma or semicolon-separated values in the input, e.g. "peanuts, gluten". Only the questions listed in LFS-1042 should have this functionality enabled.

### Questions:

1. Should this functionality be **OPT-IN** (only questions with a specific flag `enableSeparatorDetection` should have it) or **OPT-OUT** (disable where commas are needed in the values)? I propose **OPT-IN**.
2. Is the secondary color (first pic) the right choice for the UI, or is primary color (second pic) more appropriate and less alarming?

![image](https://user-images.githubusercontent.com/651980/111800930-d421da80-88a2-11eb-9ba5-b720962e9d5a.png)

![image](https://user-images.githubusercontent.com/651980/111801074-f9164d80-88a2-11eb-85fb-311118d85686.png)

marta- committed:
LFS-1041: As a user entering answers to a multi-valued question, I am warned that using "," or ";" to separate values in the input does not automatically generate separate answer entries

k-pace commented:
1. I like opt-in
2. I like the primary colour/blue - it's less aggressive

marta- committed:
LFS-1041: As a user entering answers to a multi-valued question, I am warned that using "," or ";" to separate values in the input does not automatically generate separate answer entries
Changed the accent color per review comments

marta- committed:
LFS-1041: As a user entering answers to a multi-valued question, I am warned that using "," or ";" to separate values in the input does not automatically generate separate answer entries
Made separator detection an Opt-in functionality for text questions

marta- committed:
LFS-1042: Some cardiac_rehab questions should have separator detection enabled
Baseline Health Information > Allergies > allergies

acrowthe reviewed (changes requested):
Bug found: When using the "Separate and add" option, only the last value is saved.

Steps:
1. Create a `Baseline Health Information` form
2. Into the Allergies, type in `cats, dogs; bats`
3. Click `SEPARATE AND ADD`. 3 entries will be created: `cats`, `dogs`, `bats`
4. Click save
5. Check the saved value, either in composum or by refreshing the page. Only `bats` will be saved
![image](https://user-images.githubusercontent.com/20056751/111822963-c5dfb880-88ba-11eb-9edb-dfda96bdce52.png)
![image](https://user-images.githubusercontent.com/20056751/111822975-c8421280-88ba-11eb-9bef-2221e8e1d684.png)
After a page refresh
![image](https://user-images.githubusercontent.com/20056751/111822981-c9733f80-88ba-11eb-847e-2fe266d57a48.png)

marta- committed:
LFS-1041: As a user entering answers to a multi-valued question, I am warned that using "," or ";" to separate values in the input does not automatically generate separate answer entries
Bug fix: When clicking "separate and add", only the last value is actually selected

acrowthe commented:
Tested and all values are all exported. No changes from the export format from dev, so no schema changes required

acrowthe committed:
LFS-1042: Some cardiac_rehab questions should have separator detection enabled
Enable separator detection for more cardiac rehab questions

acrowthe commented:
Implemented separator detection for all questions specified in LFS-1042, with 2 exceptions:
Skipped `Daily Actigraph Diary` >` Activity Type` > other option as it is a single selection question. `Was physical activity performed?` must be yes to display this field.
`Event - Medication Change` > `med_cat_other` also skipped as it is also a single entry question

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -270,11 +302,12 @@ function MultipleChoice(props) {
         disabled={disabled}
         onFocus={() => {maxAnswers === 1 && selectOption(ghostValue, ghostName)}}
-        onBlur={acceptEnteredOption}
+        onBlur={separatorDetected ? ()=>{} : acceptEnteredOption}
```
Hm, possible data loss here: if the user moves away and doesn't interact with the assistant, the value in the input will not be saved.

marta- replied:
Addressed in 1307151

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -291,6 +324,18 @@ function MultipleChoice(props) {
         InputProps={muiInputProps}
         inputRef={ref => {inputEl = ref}}
       />
+      { maxAnswers !== 1 &&
```
Shouldn't we also check `enableSeparatorDetection`? Without this check, all multiselect inputs have an assistant, even if not used.

marta- replied:
Done in 383442e

sdumitriu reviewed (commented):
This works, but I don't like that this is so deeply coupled with MultipleChoice. We'll have to find a cleaner way to add features like this without having to make such deep changes.

marta- committed:
LFS-1041: As a user entering answers to a multi-valued question, I am warned that using "," or ";" to separate values in the input does not automatically generate separate answer entries
Addressing review comments: avoid rendering UserInputAssistant components that will never be used

marta- committed:
LFS-1041: As a user entering answers to a multi-valued question, I am warned that using "," or ";" to separate values in the input does not automatically generate separate answer entries
Bug fix: prevent data loss when user ignores the separator detection assistant

marta- commented:
> This works, but I don't like that this is so deeply coupled with MultipleChoice. We'll have to find a cleaner way to add features like this without having to make such deep changes.

This specific feature (detecting [,;] separators in the text the user types in the input and splitting entries before saving) only makes sense for MultipleChoice questions. The only other place it might be useful is vocab questions, which should be using the MultipleChoice component anyway and will eventually be refactored in this direction. However, any concrete suggestions of decoupling and generalizing are very welcome.
